yarn-project/foundation/src/serialize/types.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * Represents a fixed-length array.
+ */
+export type TupleOf<T, N extends number> = N extends N ? (number extends N ? T[] : _TupleOf<T, N, []>) : never;
+/**
+ * Recursive type helper for constructing a fixed-length tuple of a given type.
+ * This is utilized internally by TupleOf to create the final fixed-length tuple.
+ */
+type _TupleOf<T, N extends number, R extends unknown[]> = R['length'] extends N ? R : _TupleOf<T, N, [T, ...R]>;
```

Comment:
This is all very confusing to me

Reply:
I can comment with it expands to, but it sort of is necessarily a bit magic.

TupleOf<string, 4> == [string, string, string, string]